artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/MessageReferenceImpl.java

```diff
@@ -70,9 +70,7 @@ public int compare(MessageReference o1, MessageReference o2) {
 
    private final Queue queue;
 
-   private long consumerID;
-
-   private boolean hasConsumerID = false;
+   private long consumerID = -1L;
 
```

Comment:
Are we 100% sure consumerid cannot ever be -1

Reply:
Im not sure how safe this optimisation is especially if someone implements custom id generators

Reply:
BatchingIDGenerator looks as if it can loop round into - numbers, this is used by mqtt publish manager via storage handler

Reply:
I'm not worried about our own generators since they all start from zero and it would take long time to loop round especially used for consumer id. Furthermore, number wrapping should not be allowed bcs this would resulting in the consumer id clashing. But for custom id generators, it's not safe bcs they maybe start from some negative numbers.

The purpose of this optimization is to prevent RefsOperation::getListOnConsumer from throwing exception. There is some chance that ref.emptyConsumerId() will be called Between ref.hasConsumerId() and ref.getConsumerId() == consumerID, which would throw IllegalStateException. I don't feel strong about this optimization since getListOnConsumer is triggered only by management service which would not affect core logic even if race conditions happen. I can remove it and fix it later when better idea emerges.

Reply:
Simpler change to address that is to set the has field after the id field. If we are to ensure no negative then we should ensure all if genetators down wrap into negative. Currently only the simple one enforces that. Others like the batcher do not

Reply:
For current code, when hasConsumerId returns true, thread1 will call getConsumerId. At this moment if another thread calls emptyConsumerId just before thread1 calling, thread1 would throw exception. Setting the has field after the id field can't address this.

Reply:
Would a simpler option be simply make this then Long object?